### PR TITLE
8268019: C2: assert(no_dead_loop) failed: dead loop detected

### PR DESCRIPTION
**Pull request by chhagedorn**
In the testcase, a path containing a loop is found to be dead. These nodes are then removed during IGVN after parsing (before loop opts). One of these nodes is a loop header region through which an `If` is split. Splitting an `If` through a loop header region is not intended and was disabled by [JDK-8232539](https://bugs.openjdk.java.net/browse/JDK-8232539). However, the bailouts in the current code are not enough to prevent the split if optimization for this dying loop header node due to the following reasons:
- The region is not a `LoopNode`, yet, and we do not bail out at:
https://github.com/openjdk/jdk/blob/cd8783c08ee18167f15df621e997015b971bfb01/src/hotspot/share/opto/ifnode.cpp#L119
- The loop header region has only one phi left (the other phis were already killed) and we do not bail out at:
https://github.com/openjdk/jdk/blob/cd8783c08ee18167f15df621e997015b971bfb01/src/hotspot/share/opto/ifnode.cpp#L143-L148
- The phi merges a `ConP` constant (dead path) and a `CastPP` node which is also allowed as constant:
https://github.com/openjdk/jdk/blob/cd8783c08ee18167f15df621e997015b971bfb01/src/hotspot/share/opto/ifnode.cpp#L98-L101
- The loop entry was already replaced by top and thus no predicates can be found to bail out:
https://github.com/openjdk/jdk/blob/cd8783c08ee18167f15df621e997015b971bfb01/src/hotspot/share/opto/ifnode.cpp#L244-L247

These conditions let the split if optimization to be applied without a bailout. The `CastPP` node is a use of the phi and is thus rewired in the process. The problem now is that the `CastPP` node is an input *and* output of the loop phi while being a constant to be merged:
before_split_if

This lets the split if optimization to create a data loop by setting the input of the `CastPP` to itself.

This scenario is rare and highly depends on the order in which IGVN processes the nodes. As a fix, I propose to extend the bailout checks for loop header regions to catch this edge case. The new check looks at all inputs of the region and bail out if only one path is non-top. This also avoids unnecessary splits through regions which are dying anyways.

Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268019](https://bugs.openjdk.java.net/browse/JDK-8268019): C2: assert(no_dead_loop) failed: dead loop detected

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4860/head:pull/4860` \
`$ git checkout pull/4860`

Update a local copy of the PR: \
`$ git checkout pull/4860` \
`$ git pull https://git.openjdk.java.net/jdk pull/4860/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4860`

View PR using the GUI difftool: \
`$ git pr show -t 4860`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4860.diff">https://git.openjdk.java.net/jdk/pull/4860.diff</a>

</details>

### CODE REVIEW

**Comment by chhagedorn** [comment]
Ping - due to outage

File: src/hotspot/share/opto/ifnode.cpp [hunk 1]
```diff
@@ -114,16 +114,33 @@ static Node* split_if(IfNode *iff, PhaseIterGVN *igvn) {
   if( !t->singleton() ) return NULL;
 
   // No intervening control, like a simple Call
-  Node *r = iff->in(0);
-  if( !r->is_Region() ) return NULL;
-  if (r->is_Loop()) return NULL;
-  if( phi->region() != r ) return NULL;
+  Node* r = iff->in(0);
+  if (!r->is_Region() || r->is_Loop() || phi->region() != r) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/ifnode.cpp:118]: May be also add ` || r->as_Region()->is_copy() || igvn->_worklist.member(r)` checks.

Comment (chhagedorn) [reply src/hotspot/share/opto/ifnode.cpp:118]: Could we then miss a split if opportunity when bailing out with `igvn->_worklist.member(r)`? Should we readd the `If` node to the worklist in this case? We could also think about adding `igvn->_worklist.member(r)` further down when we are really sure that we could do a split if.

Comment (vnkozlov) [reply src/hotspot/share/opto/ifnode.cpp:118]: My thinking is that Region could go away after it is processed (when input[s] are dead/top). Why bother to split through it?

I am concern that you are checking dead inputs by hand in these changes - input's transformation to Top should put Region node on worklist. Can you do experiment without your dead inputs check and only checking worklist?

I think the issue (as usual) is the order of processing of worklist. That is why I suggested this check.
Yes, you can delay worklist checking to see if we should put If node back on worklist only when we can do split if.

Comment (chhagedorn) [reply src/hotspot/share/opto/ifnode.cpp:118]: You're right, this additional worklist check should then be enough and is more clean. I reverted the fix and added a worklist check at the latest possible location. This works for repeated runs with `StressIGVN` + `RepeatCompilation`. I also included the `is_copy()` check.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

**Review by TobiHartmann — APPROVED** [review]
Nice analysis. Looks good to me.

**Comment by TobiHartmann** [comment]
And just for the record: We need to be careful to not create an infinite loop of node processing when "manually" re-adding nodes that haven't changed to the worklist. For example, while processing node A, it is re-added to the worklist because node B is still on the worklist. Then while node B is processed, it's also re-added to the worklist because node A is still on the worklist and so on. I don't think that's an issue here but I've debugged similar bugs in the past in Valhalla.

**Comment by chhagedorn** [comment]
Thank you for your review Tobias!

> And just for the record: We need to be careful to not create an infinite loop of node processing when "manually" re-adding nodes that haven't changed to the worklist. For example, while processing node A, it is re-added to the worklist because node B is still on the worklist. Then while node B is processed, it's also re-added to the worklist because node A is still on the worklist and so on. I don't think that's an issue here but I've debugged similar bugs in the past in Valhalla.

That's a good point to keep in mind. I also think it's fine in this case but in general we need to be careful when doing it.

**Pull request closed without merging**